File: 2024/src/day16/day16.py
from queue import PriorityQueue


def processing():
    i = open("src/day16/input.txt").read().splitlines()
    grid = {(r, c): v for r in range(len(i)) for c, v in enumerate(i[r])}
    return grid, len(i), len(i[0])


def dijkstra(grid, starts, rows, cols):
    frontier = PriorityQueue()
    best = dict()
    for start in starts:
        node, direction = start
        frontier.put((0, node, direction))
    while not frontier.empty():
        cost, node, direction = frontier.get()
        if (node, direction) in best and cost >= best[(node, direction)]:
            continue
        best[(node, direction)] = cost
        for dr, dc in [direction, (direction[1], -direction[0]), (-direction[1], direction[0])]:
            r, c = node[0] + dr, node[1] + dc
            if 0 <= r < rows and 0 <= c < cols and grid[(r, c)] in ['.', 'S', 'E'] and (dr, dc) == direction:
                frontier.put((cost + 1, (r, c), (dr, dc)))
            if (dr, dc) != direction:
                frontier.put((cost + 1000, node, (dr, dc)))
    return best


def part1() -> int:
    grid, rows, cols = processing()
    start, end = (rows - 2, 1), (1, cols - 2)
    paths = dijkstra(grid, [[start, (0, 1)]], rows, cols)
    return min(paths[(end, (dr, dc))] for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)])


def part2() -> int:
    grid, rows, cols = processing()
    start, end = (rows - 2, 1), (1, cols - 2)
    from_start, from_end = [dijkstra(grid, begin, rows, cols) for begin in
                            ([[start, (0, 1)]], [[end, (dr, dc)] for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)]])]
    return len({(r, c) for r in range(rows) for c in range(cols) for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)] if
                from_start.get(((r, c), (dr, dc)), 0) + from_end.get(((r, c), (-dr, -dc)), 0) ==
                min(from_start[(end, (dr, dc))] for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)])})


if __name__ == "__main__":
    print(part1())
    print(part2())

File: 2023/src/day16/day16.go
package main

import (
	"2023/src/utils"
	"fmt"
	"strings"
)

type Ray struct {
	r, c, dr, dc int
}

func processing(filename string) utils.BoundedGrid[byte] {
	lines := strings.Split(utils.ReadFile(filename), "\n")
	grid, dims := make(utils.Grid[byte]), utils.Dims{Rows: len(lines), Cols: len(lines[0])}
	for r, line := range lines {
		for c, ch := range line {
			grid[utils.Pos{R: r, C: c}] = byte(ch)
		}
	}
	return utils.BoundedGrid[byte]{Grid: grid, Dims: dims}
}

func add(seen *utils.Set[Ray], queue *utils.Queue[Ray], next Ray) {
	if !seen.Contains(next) {
		seen.Add(next)
		queue.Put(next)
	}
}

func propagate(grid utils.BoundedGrid[byte], start Ray) utils.Set[Ray] {
	seen := utils.NewSet[Ray]()
	queue := utils.NewQueue[Ray](0)
	queue.Put(start)
	for !queue.Empty() {
		current := queue.Get()
		r, c, dr, dc := current.r, current.c, current.dr, current.dc
		r, c = r+dr, c+dc
		if r < 0 || r >= grid.Dims.Rows || c < 0 || c >= grid.Dims.Cols {
			continue
		}
		ch := grid.Grid[utils.Pos{R: r, C: c}]
		switch {
		case ch == 47:
			add(&seen, &queue, Ray{r: r, c: c, dr: -dc, dc: -dr})
		case ch == 92:
			add(&seen, &queue, Ray{r: r, c: c, dr: dc, dc: dr})
		case ch == 124 && dr == 0:
			add(&seen, &queue, Ray{r: r, c: c, dr: -1, dc: 0})
			add(&seen, &queue, Ray{r: r, c: c, dr: 1, dc: 0})
		case ch == 45 && dc == 0:
			add(&seen, &queue, Ray{r: r, c: c, dr: 0, dc: -1})
			add(&seen, &queue, Ray{r: r, c: c, dr: 0, dc: 1})
		default:
			add(&seen, &queue, Ray{r: r, c: c, dr: dr, dc: dc})
		}
	}
	return seen
}

func calculate(seen utils.Set[Ray]) int {
	energy := utils.NewSet[utils.Pos]()
	for _, ray := range seen.List() {
		energy.Add(utils.Pos{R: ray.r, C: ray.c})
	}
	return energy.Size()
}

func part1(filename string) int {
	return calculate(propagate(processing(filename), Ray{r: 0, c: -1, dr: 0, dc: 1}))
}

func part2(filename string) int {
	grid := processing(filename)
	starts := make([]Ray, 0)
	for r := 0; r < grid.Dims.Rows; r++ {
		starts = append(starts, Ray{r: r, c: -1, dr: 0, dc: 1}, Ray{r: r, c: grid.Dims.Cols, dr: 0, dc: -1})
	}
	for c := 0; c < grid.Dims.Cols; c++ {
		starts = append(starts, Ray{r: -1, c: c, dr: 1, dc: 0}, Ray{r: grid.Dims.Rows, c: c, dr: -1, dc: 0})
	}
	return utils.Max(utils.Map2(starts, func(start Ray) int {
		return calculate(propagate(grid, start))
	}), nil)
}

func main() {
	filename := "day16/input.txt"
	fmt.Println(part1(filename))
	fmt.Println(part2(filename))
}
